File: src/main/java/ru/trukhmanov/twochairsbackend/service/StoreService.java
package ru.trukhmanov.twochairsbackend.service;

import jakarta.transaction.Transactional;
import org.springframework.stereotype.Service;
import ru.trukhmanov.twochairsbackend.dto.store.ProductDto;
import ru.trukhmanov.twochairsbackend.dto.store.PurchaseResponse;
import ru.trukhmanov.twochairsbackend.entity.Product;
import ru.trukhmanov.twochairsbackend.entity.UserPurchase;
import ru.trukhmanov.twochairsbackend.repository.ProductRepository;
import ru.trukhmanov.twochairsbackend.repository.UserPurchaseRepository;
import ru.trukhmanov.twochairsbackend.repository.UserRepository;

import java.time.OffsetDateTime;
import java.util.List;

@Service
public class StoreService {

    private final ProductRepository productRepository;
    private final UserPurchaseRepository purchaseRepository;
    private final UserRepository userRepository;

    public StoreService(ProductRepository productRepository,
                        UserPurchaseRepository purchaseRepository,
                        UserRepository userRepository) {
        this.productRepository = productRepository;
        this.purchaseRepository = purchaseRepository;
        this.userRepository = userRepository;
    }

    public List<ProductDto> products() {
        return productRepository.findByActiveTrueOrderByIdAsc().stream()
                .map(p -> new ProductDto(
                        p.getId(),
                        p.getType(),
                        p.getTitle(),
                        p.getPriceRub(),
                        p.getDeckId(),
                        p.isActive()))
                .toList();
    }

    @Transactional
    public PurchaseResponse purchase(long userId, long productId) {
        Product product = productRepository.findById(productId)
                .orElseThrow(() -> new IllegalArgumentException("Product not found"));

        if (!product.isActive()) throw new IllegalArgumentException("Product is not active");

        if (purchaseRepository.existsByUserIdAndProductId(userId, productId)) {
            throw new IllegalArgumentException("Already purchased");
        }

        // минимальная валидация продукта
        if ("DECK".equals(product.getType())) {
            if (product.getDeckId() == null) throw new IllegalArgumentException("Product deckId is null");
        }

        if ("PREMIUM".equals(product.getType())) {
            var user = userRepository.findById(userId).orElseThrow();
            user.setPremium(true);
        }

        UserPurchase purchase = purchaseRepository.save(UserPurchase.builder()
                .userId(userId)
                .productId(productId)
                .purchasedAt(OffsetDateTime.now())
                .build());

        return new PurchaseResponse(purchase.getId(), product.getId(), product.getType());
    }
}

File: src/main/kotlin/ru/trukhmanov/twochairsbackend/service/StoreService.kt
package ru.trukhmanov.twochairsbackend.service

import jakarta.transaction.Transactional
import org.springframework.stereotype.Service
import ru.trukhmanov.twochairsbackend.dto.store.ProductDto
import ru.trukhmanov.twochairsbackend.dto.store.PurchaseResponse
import ru.trukhmanov.twochairsbackend.entity.UserPurchase
import ru.trukhmanov.twochairsbackend.repository.ProductRepository
import ru.trukhmanov.twochairsbackend.repository.UserPurchaseRepository
import ru.trukhmanov.twochairsbackend.repository.UserRepository
import java.time.OffsetDateTime

@Service
class StoreService(
    private val productRepository: ProductRepository,
    private val purchaseRepository: UserPurchaseRepository,
    private val userRepository: UserRepository
) {

    fun products(): List<ProductDto> {
        return productRepository.findByActiveTrueOrderByIdAsc()
            .map { product ->
                ProductDto(
                    id = requireNotNull(product.id),
                    type = product.type,
                    title = product.title,
                    priceRub = product.priceRub,
                    deckId = product.deckId,
                    active = product.active
                )
            }
    }

    @Transactional
    fun purchase(userId: Long, productId: Long): PurchaseResponse {
        val product = productRepository.findById(productId)
            .orElseThrow { IllegalArgumentException("Product not found") }

        if (!product.active) {
            throw IllegalArgumentException("Product is not active")
        }

        if (purchaseRepository.existsByUserIdAndProductId(userId, productId)) {
            throw IllegalArgumentException("Already purchased")
        }

        if (product.type == "DECK" && product.deckId == null) {
            throw IllegalArgumentException("Product deckId is null")
        }

        if (product.type == "PREMIUM") {
            val user = userRepository.findById(userId).orElseThrow()
            user.premium = true
        }

        val purchase = purchaseRepository.save(
            UserPurchase(
                userId = userId,
                productId = productId,
                purchasedAt = OffsetDateTime.now()
            )
        )

        return PurchaseResponse(
            purchaseId = requireNotNull(purchase.id),
            productId = requireNotNull(product.id),
            productType = product.type
        )
    }
}
